### PR TITLE
xsdk version 0.2.0

### PR DESCRIPTION
**Pull request by BarrySmith**
A bundle that builds all the xsdk packages with consistent variants.

Includes support for version xsdk-0.2.0 and minor fixes for the various package package files.

### CODE REVIEW

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 1]
```diff
@@ -106,16 +107,26 @@ class Petsc(Package):
...
-    depends_on('superlu-dist@5.0.0:~int64', when='@for-pflotran-0.1.0+superlu-dist+mpi~int64')
-    depends_on('superlu-dist@5.0.0:+int64', when='@for-pflotran-0.1.0+superlu-dist+mpi+int64')
+    depends_on('superlu-dist@xsdk-0.2.0~int64', when='@xsdk-0.2.0+superlu-dist+mpi~int64')
+    depends_on('superlu-dist@xsdk-0.2.0+int64', when='@xsdk-0.2.0+superlu-dist+mpi+int64')
+    depends_on('superlu-dist@develop~int64', when='@develop+superlu-dist+mpi~int64')
+    depends_on('superlu-dist@develop+int64', when='@develop+superlu-dist+mpi+int64')
     depends_on('mumps+mpi', when='+mumps+mpi~int64')
     depends_on('scalapack', when='+mumps+mpi~int64')
-
+    depends_on('scalapack', when='+mumps+mpi')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: currently `mumps` is on only when `~int64`. does petsc use `scalapack` regardless of `mumps`? then you probably need another variant.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 2]
```diff
@@ -140,14 +151,10 @@ def mpi_dependent_options(self):
...
             compiler_opts = [
-                '--with-mpi=1',
-                '--with-mpi-dir=%s' % self.spec['mpi'].prefix,
+                '--with-cc=%s' % self.spec['mpi'].mpicc,
+                '--with-cxx=%s' % self.spec['mpi'].mpicxx,
+                '--with-fc=%s' % self.spec['mpi'].mpifc
             ]
-        if sys.platform != "darwin":
-            compiler_opts.extend([
-                '--with-cpp=cpp',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: @eschnett had problems on his HPC and added those lines in https://github.com/LLNL/spack/pull/1807 . Do you have some issues when `--with-cpp` is set?

File: lib/spack/spack/spec.py [hunk 3]
```diff
@@ -2650,7 +2650,7 @@ def write(s, c):
                         a_str = ' arch' + c + str(self.architecture) + ' '
                         write(fmt % (a_str), c)
                 elif c == '/':
-                    out.write('/' + fmt % (self.dag_hash(7)))
+                    out.write('-' + fmt % (self.dag_hash(7)))
```
Comment (adamjstewart) [inline lib/spack/spack/spec.py]: This is a more controversial change and should probably go in a different PR.

File: var/spack/repos/builtin/packages/alquimia/package.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    depends_on('hdf5')
+    depends_on('pflotran@xsdk-0.2.0', when='@xsdk-0.2.0')
+    depends_on('pflotran@develop', when='@develop')    
+    depends_on('petsc@xsdk-0.2.0', when='@xsdk-0.2.0')
+    depends_on('petsc@develop', when='@develop')    
+
+    def cmake_args(self):
+        spec = self.spec
+
+        options = ['-DCMAKE_C_COMPILER=%s' % self.spec['mpi'].mpicc,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/alquimia/package.py]: you may as well join this with the next lines as
```
options = [
   '-DCMAKE_....',
   ...
]
```
also you set `spec= self.spec` above, but don't use it right here.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 5]
```diff
@@ -140,14 +151,10 @@ def mpi_dependent_options(self):
                 raise RuntimeError('\n'.join(errors))
         else:
             compiler_opts = [
-                '--with-mpi=1',
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: does `petsc` not need this to understand that the provided compilers are MPI wrappers?

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 6]
```diff
@@ -112,23 +115,28 @@ class Trilinos(CMakePackage):
     depends_on('scalapack', when='+mumps')
     depends_on('superlu-dist@:4.3', when='@:12.6.1+superlu-dist')
     depends_on('superlu-dist', when='@12.6.2:+superlu-dist')
+    depends_on('superlu-dist@develop', when='@develop:+superlu-dist')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/trilinos/package.py]: you probably don't need `:` in `@develop:`.

File: var/spack/repos/builtin/packages/trilinos/package.py [hunk 7]
```diff
@@ -183,6 +191,11 @@ def cmake_args(self):
         else:
             options.append('-DCMAKE_INSTALL_NAME_DIR:PATH=%s/lib' % prefix)
 
+        if '+tpetra' in spec:
+            options.append('-DTrilinos_ENABLE_Tpetra:BOOL=ON')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/trilinos/package.py]: this could be a one-liner above in the main list:
```
'-DTrilinos_ENABLE_Tpetra:BOOL=%s' % (
   'ON' if '+tpetra' in spec else 'OFF'),
```

File: var/spack/repos/builtin/packages/xsdktrilinos/package.py [hunk 8]
```diff
@@ -49,11 +50,14 @@ class Xsdktrilinos(CMakePackage):
 
     # MPI related dependencies
     depends_on('mpi')
-    depends_on('hypre~internal-superlu', when='+hypre')
```
Comment (davydden) [inline var/spack/repos/builtin/packages/xsdktrilinos/package.py]: so for `12.8.1` you don't need `hypre` anymore?

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 5 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/petsc/package.py]: Nope, --with-mpi is the default.

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 2 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/petsc/package.py]: Yes, the "cpp" passed here may not correspond to the compiler that is used; producing inconsistencies. 

I believe that the problem that required the '--with-cpp=cpp', has been fixed.

File: var/spack/repos/builtin/packages/alquimia/package.py (hunk 4 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/alquimia/package.py]: Done

File: var/spack/repos/builtin/packages/petsc/package.py (hunk 1 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/petsc/package.py]: Fixed

File: var/spack/repos/builtin/packages/trilinos/package.py (hunk 6 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Fixed

File: var/spack/repos/builtin/packages/trilinos/package.py (hunk 7 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/trilinos/package.py]: Done

File: var/spack/repos/builtin/packages/xsdktrilinos/package.py (hunk 8 above)
Comment (BarrySmith) [inline var/spack/repos/builtin/packages/xsdktrilinos/package.py]: Fixed

File: lib/spack/spack/spec.py (hunk 3 above)
Comment (BarrySmith) [inline lib/spack/spack/spec.py]: I have removed this change since it may no longer be needed. If it a test determines later it is needed then it will be changed in a separate pull request.

File: var/spack/repos/builtin/packages/xsdk/package.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    depends_on('alquimia@xsdk-0.2.0', when='@xsdk-0.2.0')
+    depends_on('alquimia@develop', when='@develop')    
+
+    depends_on('xsdktrilinos@xsdk-0.2.0', when='@xsdk-0.2.0')
+    depends_on('xsdktrilinos@develop', when='@develop')    
+
+    variant('debug', default=False, description='Compile in debug mode')
+
+    # How do we propagate debug flag to all depends on packages ?
+    # If I just do spack install xsdk+debug will that propogate it down?
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/xsdk/package.py]: There is currently no way to do this. It won't do it automatically. @citibeth has been arguing for global variants and I've been arguing for automatic propagation of variants through the concretizer. Both probably aren't going to happen in the near future.

File: var/spack/repos/builtin/packages/petsc/package.py [hunk 10]
```diff
@@ -66,11 +65,13 @@ class Petsc(Package):
     variant('boost',   default=True,  description='Activates support for Boost')
     variant('hypre',   default=True,
             description='Activates support for Hypre (only parallel)')
-    variant('mumps',   default=True,
```
Comment (davydden) [inline var/spack/repos/builtin/packages/petsc/package.py]: could you please add a comment above this line why it's disabled by default so that others won't try to enable it when modifying `petsc` package next time.

Comment (BarrySmith) [inline var/spack/repos/builtin/packages/petsc/package.py]: Done

**Comment by BarrySmith** [comment]
@tgamblin can we get this accepted soon?

**Comment by adamjstewart** [comment]
Just need to fix a few more Flake8 errors and I'll merge this:
```
var/spack/repos/builtin/packages/xsdk/package.py:51: [E501] line too long (99 > 79 characters)
var/spack/repos/builtin/packages/xsdk/package.py:53: [E501] line too long (106 > 79 characters)
var/spack/repos/builtin/packages/xsdk/package.py:56: [E501] line too long (89 > 79 characters)
var/spack/repos/builtin/packages/xsdk/package.py:58: [E501] line too long (86 > 79 characters)
var/spack/repos/builtin/packages/xsdk/package.py:69: [E265] block comment should start with '# '
var/spack/repos/builtin/packages/xsdk/package.py:70: [E265] block comment should start with '# '
```

**Comment by tgamblin** [comment]
There are some spurious flake8 errors in here.  I merged #3932 but the long `depends_on()` lines in `xsdk/package.py` should be excepted and they're not.  I'll fix it.

**Comment by tgamblin** [comment]
Tests should pass now that #3990 is merged.  I'll merge this once that is done.
